File: backend/app/models/surf.py
from __future__ import annotations
from pydantic import BaseModel, Field
from typing import Optional, Literal
from datetime import datetime


class BuoyReading(BaseModel):
    station_id: str
    timestamp: datetime
    wvht_m: Optional[float] = None      # significant wave height (meters)
    wvht_ft: Optional[float] = None     # derived: wvht_m * 3.28084
    dpd_s: Optional[float] = None       # dominant period (seconds)
    apd_s: Optional[float] = None       # average period (seconds)
    mwd_deg: Optional[float] = None     # mean wave direction (degrees true)
    mwd_label: Optional[str] = None     # e.g. "SSW"
    wspd_ms: Optional[float] = None     # wind speed (m/s)
    wspd_mph: Optional[float] = None    # derived
    wdir_deg: Optional[float] = None    # wind direction (degrees true)
    wdir_label: Optional[str] = None    # e.g. "WNW"
    wtmp_c: Optional[float] = None      # water temp (Celsius)
    wtmp_f: Optional[float] = None      # derived
    atmp_c: Optional[float] = None      # air temp (Celsius)
    pres_hpa: Optional[float] = None    # pressure (hPa)
    data_age_minutes: Optional[float] = None


WaveClassification = Literal[
    "flat", "ankle", "knee", "waist", "chest", "head", "overhead", "double-overhead", "XXL"
]


class WavePower(BaseModel):
    kw_per_meter: float
    classification: WaveClassification
    surf_rating: int = Field(ge=0, le=10)


class WindQuality(BaseModel):
    direction_deg: float
    direction_label: str
    speed_mph: float
    quality: Literal["offshore", "cross-offshore", "cross", "cross-onshore", "onshore"]
    quality_label: str


class SpotMeta(BaseModel):
    id: str
    name: str
    short_name: str
    lat: float
    lon: float
    break_type: str
    difficulty: str
    facing_dir: str
    region: str
    description: str
    fame_score: float


class BreakingConditions(BaseModel):
    buoy_hs_ft: float
    buoy_period_s: float
    buoy_dir_deg: float
    swell_type: str
    swell_type_short: str
    period_quality: str
    swell_angle_diff: float
    direction_rating: str
    direction_pct: float
    breaking_hs_ft: float
    face_height_min_ft: float
    face_height_max_ft: float
    face_height_label: str
    interpretation: str
    spot_context: str


class SunTimes(BaseModel):
    date: str
    first_light: str
    first_light_display: str
    sunrise: str
    sunrise_display: str
    solar_noon_display: str
    sunset: str
    sunset_display: str
    last_light: str
    last_light_display: str
    dawn_patrol_window: dict
    daylight_hours: float
    minutes_to_sunrise: int
    minutes_to_sunset: int
    is_daytime: bool
    is_dawn_patrol_window: bool
    is_golden_hour_morning: bool


class SurfCondition(BaseModel):
    spot_id: str
    spot_name: str
    spot_short_name: str
    updated_at: datetime
    buoy: BuoyReading
    wave_power: Optional[WavePower] = None
    breaking: Optional[BreakingConditions] = None
    wind: Optional[WindQuality] = None
    crowd: Optional["CrowdPrediction"] = None
    next_tide: Optional["TideEvent"] = None
    sun: Optional[SunTimes] = None


# avoid circular import — imported from crowd.py and tide.py at runtime
from app.models.crowd import CrowdPrediction  # noqa: E402
from app.models.tide import TideEvent          # noqa: E402

SurfCondition.model_rebuild()

File: frontend/src/types/surf.ts
export interface BuoyReading {
  station_id: string
  timestamp: string
  wvht_m: number | null
  wvht_ft: number | null
  dpd_s: number | null
  apd_s: number | null
  mwd_deg: number | null
  mwd_label: string | null
  wspd_ms: number | null
  wspd_mph: number | null
  wdir_deg: number | null
  wdir_label: string | null
  wtmp_c: number | null
  wtmp_f: number | null
  atmp_c: number | null
  pres_hpa: number | null
  data_age_minutes: number | null
}

export type WaveClassification =
  | 'flat' | 'ankle' | 'knee' | 'waist' | 'chest'
  | 'head' | 'overhead' | 'double-overhead' | 'XXL'

export interface WavePower {
  kw_per_meter: number
  classification: WaveClassification
  surf_rating: number
}

export type WindQualityType = 'offshore' | 'cross-offshore' | 'cross' | 'cross-onshore' | 'onshore'

export interface WindQuality {
  direction_deg: number
  direction_label: string
  speed_mph: number
  quality: WindQualityType
  quality_label: string
}

export type CrowdLevel = 'empty' | 'uncrowded' | 'moderate' | 'crowded' | 'packed'

export interface CrowdPrediction {
  score: number
  level: CrowdLevel
  confidence: number
  peak_hour_today: number | null
}

export interface TideEvent {
  event_type: 'high' | 'low'
  timestamp: string
  height_ft: number
  hours_away: number | null
}

export interface TidePrediction {
  timestamp: string
  height_ft: number
}

export interface TideWindow {
  station_id: string
  generated_at: string
  predictions: TidePrediction[]
  events: TideEvent[]
}

export interface BreakingConditions {
  buoy_hs_ft: number
  buoy_period_s: number
  buoy_dir_deg: number
  swell_type: string
  swell_type_short: string
  period_quality: 'excellent' | 'good' | 'fair' | 'poor'
  swell_angle_diff: number
  direction_rating: 'ideal' | 'good' | 'cross' | 'marginal' | 'blocked'
  direction_pct: number
  breaking_hs_ft: number
  face_height_min_ft: number
  face_height_max_ft: number
  face_height_label: string
  interpretation: string
  spot_context: string
}


export interface SunTimes {
  date: string
  first_light_display: string
  sunrise_display: string
  solar_noon_display: string
  sunset_display: string
  last_light_display: string
  minutes_to_sunrise: number
  minutes_to_sunset: number
  is_daytime: boolean
  is_dawn_patrol_window: boolean
  is_golden_hour_morning: boolean
}

export interface SwellComponent {
  label: string          // "Primary", "Secondary", "Third", "Wind Chop"
  height_ft: number
  period_s: number
  direction_deg: number
  direction_label: string
}

export interface SurfCondition {
  spot_id: string
  spot_name: string
  spot_short_name: string
  updated_at: string
  buoy: BuoyReading
  wave_power: WavePower | null
  breaking: BreakingConditions | null
  wind: WindQuality | null
  crowd: CrowdPrediction | null
  next_tide: TideEvent | null
  current_tide_ft: number | null
  sun: SunTimes | null
  swells: SwellComponent[]
}

export interface SpotMeta {
  id: string
  name: string
  short_name: string
  lat: number
  lon: number
  break_type: string
  difficulty: string
  facing_dir: string
  region: string
  description: string
  fame_score: number
}

export interface ForecastHour {
  timestamp: string
  date: string
  hour: number
  face_height_min_ft: number
  face_height_max_ft: number
  face_height_label: string
  swell_height_m: number | null
  swell_period_s: number | null
  swell_dir_deg: number | null
  swell_type_short: string
  period_quality: 'excellent' | 'good' | 'fair' | 'poor'
  direction_rating: string
  direction_pct: number
  wind_speed_mph: number
  wind_quality: WindQualityType
  wind_quality_label: string
  tide_ft: number | null
  combo_swell: boolean
  combo_label: string
  surf_rating: number
  crowd_score: number
  crowd_level: CrowdLevel
}

export interface ForecastDay {
  date: string
  day_label: string
  face_height_min_ft: number
  face_height_max_ft: number
  face_height_label: string
  peak_face_ft: number
  peak_hour: number
  swell_height_m: number | null
  swell_period_s: number | null
  swell_dir_label: string
  swell_type_short: string
  period_quality: 'excellent' | 'good' | 'fair' | 'poor'
  direction_rating: string
  best_wind_quality: WindQualityType
  avg_wind_speed_mph: number
  surf_rating: number
  crowd_score: number
  crowd_level: CrowdLevel
}

export interface SpotForecast {
  spot_id: string
  spot_name: string
  generated_at: string
  daily: ForecastDay[]
  hourly: ForecastHour[]
}

export interface UserPreferences {
  min_wave_height_ft: number
  max_wave_height_ft: number
  preferred_period_s: number
  max_wind_speed_mph: number
  max_crowd_score: number
  experience_level: 'beginner' | 'intermediate' | 'advanced' | 'expert'
}

export interface SpotWindow {
  spot_id: string
  spot_name: string
  start: string
  end: string
  wave_height_ft: number
  period_s: number
  crowd_score: number
  wind_quality: string
  composite_score: number
  why: string
}

export interface AIRankingResponse {
  ranked_windows: SpotWindow[]
  explanation: string
  top_pick: string
  top_pick_time: string
  generated_at: string
  model_used: string
}
